modules/net-vpc-firewall/main.tf

```diff
@@ -15,149 +15,107 @@
+        targets              = try(rule.targets, null)
+        use_service_accounts = try(rule.use_service_accounts, false)
+      }
     ]
   ])
-
   _factory_rules = {
-    for d in local._factory_rules_raw : d["name"] => d
-  }
-
-  custom_rules = merge(local._custom_rules, local._factory_rules)
-}
-
-
-###############################################################################
-#                            rules based on IP ranges
-###############################################################################
-
-resource "google_compute_firewall" "allow-admins" {
-  count         = length(var.admin_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-admins"
-  description   = "Access from the admin subnet to all subnets"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.admin_ranges
-  allow { protocol = "all" }
-}
-
-###############################################################################
-#                              rules based on tags
-###############################################################################
-
-resource "google_compute_firewall" "allow-tag-ssh" {
-  count         = length(var.ssh_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-ssh"
-  description   = "Allow SSH to machines with the 'ssh' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.ssh_source_ranges
-  target_tags   = ["ssh"]
-  allow {
-    protocol = "tcp"
-    ports    = ["22"]
-  }
-}
-
-resource "google_compute_firewall" "allow-tag-http" {
-  count         = length(var.http_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-http"
-  description   = "Allow HTTP to machines with the 'http-server' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.http_source_ranges
-  target_tags   = ["http-server"]
-  allow {
-    protocol = "tcp"
-    ports    = ["80"]
+    for r in local._factory_rule_list : r.name => r
   }
-}
-
-resource "google_compute_firewall" "allow-tag-https" {
-  count         = length(var.https_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-https"
-  description   = "Allow HTTPS to machines with the 'https' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.https_source_ranges
-  target_tags   = ["https-server"]
-  allow {
-    protocol = "tcp"
-    ports    = ["443"]
+  _named_ranges = merge(
+    var.named_ranges,
+    try(yamldecode(file(var.factories_config.cidr_tpl_file)), {})
+  )
+  # convert rules data to resource format and replace range template variables
+  rules = {
+    for name, rule in merge(local._factory_rules, var.custom_rules) :
+    name => merge(rule, {
+      action    = rule.is_deny == true ? "DENY" : "ALLOW"
+      direction = rule.is_egress == true ? "EGRESS" : "INGRESS"
+      rules     = { for k, v in rule.rules : k => v }
```

Comment:
would `rules = rules.rules` work?

Reply:
not really as this maintains ordering, not doing this would use lexical ordering potentially triggering recreation if a rule is added to the end